test/fshtypes/ParamRuleSet.test.ts

```diff
@@ -0,0 +1,42 @@
+import 'jest-extended';
+import { ParamRuleSet } from '../../src/fshtypes/ParamRuleSet';
+import { EOL } from 'os';
+
+describe('ParamRuleSet', () => {
+  describe('#constructor', () => {
+    it('should set initial properties correctly', () => {
+      const ruleSet = new ParamRuleSet('MyParamRuleSet');
+      expect(ruleSet.name).toBe('MyParamRuleSet');
+      expect(ruleSet.parameters).toEqual([]);
+      expect(ruleSet.contents).toBe('');
+    });
+  });
+
+  describe('#applyParameters', () => {
+    it('should replace parameters with their value in the contents', () => {
+      const ruleSet = new ParamRuleSet('MyParamRuleSet');
+      ruleSet.parameters = ['first', 'second'];
+      ruleSet.contents = [
+        '* category {first}..*',
+        '* category from {second}',
+        '* note 0..{first}'
+      ].join(EOL);
+
+      const appliedContents = ruleSet.applyParameters(['5', 'MySystem']);
+      const expectedContents = ['* category 5..*', '* category from MySystem', '* note 0..5'].join(
+        EOL
+      );
+      expect(appliedContents).toBe(expectedContents);
+    });
+
+    it('should replace parameters sequentially', () => {
+      const ruleSet = new ParamRuleSet('MyParamRuleSet');
+      ruleSet.parameters = ['first', 'second'];
+      ruleSet.contents = ['* code from {first}', '* category from {second}'].join(EOL);
+
+      const appliedContents = ruleSet.applyParameters(['{second}Deluxe', 'MySystem']);
+      const expectedContents = ['* code from MySystemDeluxe', '* category from MySystem'].join(EOL);
```

Comment:
Oh... I see now.  So `['{second}Deluxe', 'MySystem']` works, but `['MySystem', '{first}Deluxe']` would not...

Hmm...  I'm not sure I like that.  I feel like either _no_ 2nd-level substitutions should happen or _all_ 2nd-level substitutions should happen.  This inbetween state is likely to confuse users.

Reply:
That's reasonable. I think that between those two options, I would want no extra substitutions to take place. I'll make that change.

Reply:
Yeah.  I think that makes sense-- because one big issue w/ recursive substitutions is that suddenly the var names are _meaningful_ outside of the RuleSet definition. So if you simply renamed a var in your RuleSet, you could break code that uses it (if we allowed recursive substitution).